tensorflow/compiler/xla/service/gpu/fusion_bitcast_lift_test.cc

```diff
@@ -0,0 +1,215 @@
+}
+)";
+  auto module = ParseAndReturnVerifiedModule(hlo_text).ValueOrDie();
+  EXPECT_TRUE(FusionBitcastLift().Run(module.get()).ValueOrDie());
+
+  auto* root = module->entry_computation()->root_instruction();
+  EXPECT_EQ(HloOpcode::kFusion, root->opcode());
+
+  // The fusion should have 1 input and it should be a bitcast with 2d output.
+  EXPECT_EQ(1, root->operands().size());
+  EXPECT_EQ(HloOpcode::kBitcast, root->operand(0)->opcode());
+  EXPECT_EQ(2, root->operand(0)->shape().rank());
+
+  // No bitcast should be left inside the fusion.
+  for (HloInstruction* instr : root->fused_instructions()) {
+    EXPECT_NE(HloOpcode::kBitcast, instr->opcode());
+  }
+
+  EXPECT_TRUE(RunAndCompare(hlo_text, ErrorSpec{1e-5, 1e-5}));
+}
+
+// Tests that we lift bitcast outside the fusion when scalar broadcasting are
+// present.
+//
+// Input graph:
+// Fusion 1x4d and 1x0d inputs, 2 1d output
+//
+// After optimization, the graph is:
+// Bitcast 4d -> 2d
+//   |
+// Fusion 1x2d and 1x0d inputs, 2 1d output
+//   Inside the fusion, there is a bitcast left after the broadcast.
+TEST_F(FusionBitcastLiftTest, ScalarBroadcastTest) {
+  const char* hlo_text = R"(
+HloModule mod
+
+%scalar_add_computation (scalar_lhs.1: f32[], scalar_rhs.1: f32[]) -> f32[] {
+  %scalar_lhs.1 = f32[] parameter(0)
+  %scalar_rhs.1 = f32[] parameter(1)
+  ROOT %add.5 = f32[] add(f32[] %scalar_lhs.1, f32[] %scalar_rhs.1)
+}
+
+%fused_computation.21_4d (param_0.59: f16[2,14,14,672], param_1: f32[]) -> (f32[672], f32[672]) {
+  %param_0.59 = f16[2,14,14,672] parameter(0)
+  %convert.21 = f32[2,14,14,672] convert(%param_0.59)
+  %bitcast.25 = f32[392,672] bitcast(%convert.21)
+  %constant_84 = f32[] constant(0)
+  %reduce.12 = f32[672]{0} reduce(%bitcast.25, %constant_84), dimensions={0}, to_apply=%scalar_add_computation
+  %param_1 = f32[] parameter(1)
+  %broadcast = f32[2,14,14,672] broadcast(%param_1), dimensions={}
+  %multiply.43.clone.1 = f32[2,14,14,672] multiply(%convert.21, %broadcast)
+  %bitcast.24.clone.1 = f32[392,672] bitcast(%multiply.43.clone.1)
+  %reduce.11.clone.1 = f32[672]{0} reduce(%bitcast.24.clone.1, %constant_84), dimensions={0}, to_apply=%scalar_add_computation
+  ROOT %tuple.13 = (f32[672]{0}, f32[672]{0}) tuple(%reduce.12, %reduce.11.clone.1)
+}
+
+ENTRY main {
+  %param_0.59 = f16[2,14,14,672] parameter(0)
+  %param_1 = f32[] parameter(1)
+  ROOT %fusion.21_4d = (f32[672]{0}, f32[672]{0}) fusion(%param_0.59, %param_1), kind=kInput, calls=%fused_computation.21_4d
+}
+)";
+  auto module = ParseAndReturnVerifiedModule(hlo_text).ValueOrDie();
+  EXPECT_TRUE(FusionBitcastLift().Run(module.get()).ValueOrDie());
+
+  auto* root = module->entry_computation()->root_instruction();
+  EXPECT_EQ(HloOpcode::kFusion, root->opcode());
+
+  // The fusion should have 2 inputs and the first one should be a
+  // bitcast with 2d output.
+  EXPECT_EQ(2, root->operands().size());
```

Comment:
Is it possible to compare the HLO after, I think it would be more readable this way?

Reply:
I checked a few tests file now and I didn't see this pattern of test being done.
Also, doing a pure string comparison would be brittle.

So we would need to use something like the llvm `opt --passname` tool. `hlo_to_llvm_ir` isn't what we want.
Also, it would miss the execution comparison to be sure that the values doesn't changes. Comparing only the code is less verification then also checking that the value is the right one.

Having those 2 extra tools would be great. To my knowledge, they do not exist. If they exist, tell me and I'll use them.

Reply:
What about the FIleCheck-based tests, like in e.g. gemm_rewriter_test?

Execution comparison can be done separately, like in that file.

Reply:
Could be done line that. But now you start to have the tests for only 1 file into multiple file. I like FileCheck based tests. If I could verify the results with that, I would use only that. But that isn't supported for now.

I can change the test if you prefer to have the tests separate. But I find that less readable then having all the tests in one system in one file.

Reply:
> But now you start to have the tests for only 1 file into multiple file

why? In gemm_rewrite_test they are all in a single file?

Reply:
I didn't know about the MatchOptimizedHlo fct. Only about the FileCheck test in separate file. I can update the test to use MatchOptimizedHlo. But it will probably be Monday.